### PR TITLE
[cluster-autoscaler][clusterapi] Remove internal types in favor of unstructured

### PR DESCRIPTION
**Pull request by detiber**
Currently we define internal types for use with Cluster API and are serializing/deserializing to/from these internal types as we interact with Cluster API resources. I believe this causes confusion since the internal types do not necessarily represent the external Cluster API types that we are interacting with (that may span multiple versions). It also makes adding support for fields that may or may not exist for a subset of Cluster API versions more difficult to maintain (since we are attempting to abstract these differences away as part of the serialization/deserialization process). 

This PR removes the use of internal type representations all together in favor of Unstructured directly.

### CODE REVIEW

**Comment by elmiko** [comment]
Jason, this looks cool. i'm still reviewing and testing, but i'm amazed at how much we can remove switching to the unstructured. nicely done!

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_provider.go [hunk 1]
```diff
@@ -142,22 +146,32 @@ func BuildClusterAPI(opts config.AutoscalingOptions, do cloudprovider.NodeGroupD
 	}
 
 	// Grab a dynamic interface that we can create informers from
-	dc, err := dynamic.NewForConfig(externalConfig)
+	managementClient, err := dynamic.NewForConfig(externalConfig)
```
Comment (elmiko) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_provider.go:149]: i thought we were separating this work into the follow up commit?

**Comment by elmiko** [comment]
just a follow up, i am seeing this build failure on master as well. so either i have a big problem or the repo does.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_provider.go (hunk 1 above)
Comment (detiber) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_provider.go:149]: In this instance I just renamed the variable to be more descriptive of it's use, there is no separation between kubeconfigs that are used by the clients with this change.

Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_provider.go:149]: ok, cool. i figured that was the case but wanted to be sure.

**Comment by detiber** [comment]
> i was trying to get this change built so i could test on a local cluster but i keep this hitting this:
> 
> ```
> $ make clean build
> rm -f cluster-autoscaler
> CGO_ENABLED=0 GO111MODULE=off GOOS=linux go build --ldflags "-s"  ./...
> # k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/component-base/cli/flag
> vendor/k8s.io/component-base/cli/flag/ciphersuites_flag.go:51:51: undefined: tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
> vendor/k8s.io/component-base/cli/flag/ciphersuites_flag.go:52:51: undefined: tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
> make: *** [Makefile:25: build] Error 2
> ```
> 
> i tried to clean extra files, and also looked at rebuilding the vendor dir, any advice?

Interesting, I'm not seeing any build errors locally, have you tried using the `build-in-docker` make target? I'm not sure if it's directly related, but I've seen similar cipher suite mismatch errors when trying to build k/k using the wrong version of Go.

**Comment by elmiko** [comment]
> Interesting, I'm not seeing any build errors locally, have you tried using the `build-in-docker` make target? I'm not sure if it's directly related, but I've seen similar cipher suite mismatch errors when trying to build k/k using the wrong version of Go.

i haven't tried the build in docker yet, it was only a few commits ago that things broke. i was just trying to rebase your stuff on top of the old commit. probably easier to just build in docker lol

**Comment by elmiko** [comment]
ok, got my build working.

i tested this against a CAPD cluster, i tried out scaling up and down with a workload and deleting the workload, also tested adding a new MachineDeployment and scaling that.

looks good so far, i didn't hit any errors with operation of the autoscaler.

/area provider/cluster-api
/lgtm

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go [hunk 2]
```diff
@@ -203,9 +178,42 @@ func (c *machineController) run(stopCh <-chan struct{}) error {
...
+	if c.machineDeploymentsAvailable {
+		machineDeployment, err := c.findMachineSetOwner(scalableResource)
+		if err != nil {
+			return nil, err
+		}
+
+		// If a machineDeployment was found return a new nodegroup for the machineDeployment
+		if machineDeployment != nil {
+			scalableResource = machineDeployment
+		}
```
[the remark this answers is not in the document]
Comment (detiber) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go]: Indeed, this also made me realize that I could make the code a bit more clear as well

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go [hunk 3]
```diff
@@ -526,102 +684,151 @@ func TestControllerFindNodeByNodeName(t *testing.T) {
...
+		namespace := RandomString(6)
+		testConfig1 := createMachineDeploymentTestConfig(namespace, RandomString(6), 5, map[string]string{
+			nodeGroupMinSizeAnnotationKey: "1",
+			nodeGroupMaxSizeAnnotationKey: "10",
+		})
+
+		// Construct a second set of objects and add the machines,
+		// nodes and the additional machineset to the existing set of
+		// test objects in the controller. This gives us two
+		// machinesets, each with their own machines and linked nodes.
```
[the remark this answers is not in the document]
Comment (detiber) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go]: Yes, thank you. Fixed with the latest changes

**Comment by JoelSpeed** [comment]
Looks good, thanks!

/lgtm

**Comment by enxebre** [comment]
thanks a lot @detiber! This looks great. I'm planning to have a deeper look at this on Monday.
Since this is a big refactor and kube 1.19 should be code frozen by now and given the unfortunate lack of e2e testing we have atm for this provider I'd suggest we hold the merge until the ca main branch opens for 1.20 and proceed with any backport if we need so. Does it make sense to you?

**Comment by detiber** [comment]
@enxebre I would like to avoid holding off until 1.20, only because this helps enable the other functionality for running against workload clusters that exist in a separate management controller vs only against self-managed clusters. That said, I will defer to the larger community to make that decision.

**Comment by detiber** [comment]
@enxebre now that the 1.19 release has been cut, any objections to proceeding with this?

**Comment by enxebre** [comment]
@MaciekPytel is master targeting 1.20 now?
thanks @detiber!
/approve
PTAL @JoelSpeed @elmiko

**Comment by detiber** [comment]
> @MaciekPytel is the main branch targeting 1.20 now?
I'm not Maciek, but yes the main branch is targeting 1.20, the `cluster-autoscaler-release-1.19` branch has already be cut along with the 1.19.0 release.

**Comment by detiber** [comment]
@MaciekPytel this should be good to go now, need cluster-autoscaler approval for the vendor changes

/assign @MaciekPytel

**Comment by elmiko** [comment]
thanks Jason!

**Comment by ncdc** [comment]
👋 hi, are there any updates on getting this reviewed for approval? Thanks!

**Comment by elmiko** [comment]
i think we are good on this from the CAPI side, just need an autoscaler owner to give approval. cc @mwielgus @MaciekPytel

**Comment by ncdc** [comment]
@elmiko it doesn't look like we need root OWNERS approval, just cluster-autoscaler. Do you think we should ask someone in https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/OWNERS?

**Comment by elmiko** [comment]
> @elmiko it doesn't look like we need root OWNERS approval, just cluster-autoscaler. Do you think we should ask someone in https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/OWNERS?

++, thanks for pointing that out Andy =)

cc @aleksandra-malinowska @feiskyer @vivekbagade

**Pull request merged**